src/nomad_simulations/properties/band_structure.py

```diff
@@ -0,0 +1,370 @@
+    ) -> Tuple[Optional[pint.Quantity], Optional[np.ndarray]]:
+        """
+        Order the eigenvalues based on the `value` and `occupation`. The return `value` and
+        `occupation` are flattened.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (Tuple[Optional[pint.Quantity], Optional[list]]): The flattened and sorted `value` and `occupation`.
+        """
+
+        # Check if `value` exists
+        if self.value is None or len(self.value) == 0:
+            logger.error('Could not find `value` defined.')
+            return None, None
+
+        # Check if `occupation` exists and have the same shape as `value`
+        if not self._check_occupation_shape(logger):
+            return None, None
+        total_shape = np.prod(self.value.shape)
+
+        # Order the indices in the flattened list of `value`
+        flattened_value = self.value.reshape(total_shape)
+        flattened_occupation = self.occupation.reshape(total_shape)
+        sorted_indices = np.argsort(flattened_value, axis=0)
+
+        sorted_value = (
+            np.take_along_axis(flattened_value.magnitude, sorted_indices, axis=0)
+            * flattened_value.u
+        )
+        sorted_occupation = np.take_along_axis(
+            flattened_occupation, sorted_indices, axis=0
+        )
+        self.m_cache['sorted_eigenvalues'] = True
+        return sorted_value, sorted_occupation
+
+    def resolve_homo_lumo_eigenvalues(
+        self, logger: BoundLogger
+    ) -> Tuple[Optional[pint.Quantity], Optional[pint.Quantity]]:
+        """
+        Resolve the `highest_occupied` and `lowest_unoccupied` eigenvalues by performing a binary search on the
+        flattened and sorted `value` and `occupation`. If these quantities already exist, overwrite them or return
+        them if it is not possible to resolve from `value` and `occupation`.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (Tuple[Optional[pint.Quantity], Optional[pint.Quantity]]): The `highest_occupied` and
+            `lowest_unoccupied` eigenvalues.
+        """
+        # Sorting `value` and `occupation`
+        sorted_value, sorted_occupation = self.order_eigenvalues(logger)
+        if sorted_value is None or sorted_occupation is None:
+            if self.highest_occupied is not None and self.lowest_unoccupied is not None:
+                return self.highest_occupied, self.lowest_unoccupied
+            return None, None
+        sorted_value_unit = sorted_value.u
+        sorted_value = sorted_value.magnitude
+
+        # Binary search ot find the transition point between `occupation = 2` and `occupation = 0`
+        tolerance = 1e-3  # TODO add tolerance from config fields
+        homo = self.highest_occupied
+        lumo = self.lowest_unoccupied
+        low_occ = 0
+        high_unocc = sorted_value.shape[-1] - 1
+        while low_occ <= high_unocc:
```

Comment:
can't you use np.argwhere(sorted_occupation > tol) here? also, how about spin-pol case?

Reply:
Hehe, I'll argue that binary search (which is logN) is more optimal than argwhere (which is linear). These arrays are going to be potentially large, but I can change it to argwhere if even tho you prefer readability.

Reply:
there is also this np.argmax that seems to return the first element matching criterion.

Reply:
It seems to be linear as well. ChatGPT helped me check this (for a sample of 10000 points)

![time_complexity](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/assets/112697669/993695ba-e2b7-42a4-8de3-3e9b34a1ee2b)

Reply:
interesting. thanks for testing. then go for the optimized approach.

Reply:
I just found `numpy.searchsorted` which seems to perform binary search. Let me try it.

Reply:
It seems to be almost the same. I tried some random arrays (hence the bit of randomness between results), averaged, and increased the size to 1e6.

![time_complexity](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/assets/112697669/802045a0-95e4-4598-bf98-d4b4cd2720aa)

Well, I can also use numpy.searchsorted if it is more readable, I don't mind actually.

Reply:
I suggest you use numpy for easy maintenance.